### PR TITLE
Copy nullable state of members unconditionally up to a certain depth

### LINKED ISSUE #31395: Nullability of class fields are not inherited on assignment
A warning is reported for `d.F`. (Compare with `t.F`.)
```c#
#nullable enable

class C
{
    internal object? F;
}

struct S
{
    internal object? F;
}

class Program
{
    static void F1()
    {
        var c = new C() { F = "" };
        F(c.F); // ok
        var d = c;
        F(d.F); // warning
    }

    static void F2()
    {
        var s = new S() { F = "" };
        F(s.F); // ok
        var t = s;
        F(t.F); // ok
    }

    static void F(object o)
    {
    }
}
```

### PR DESCRIPTION
**Pull request by cston**
Copy nullable state of members unconditionally up to a certain depth.

From a suggestion from @gafter.

The change allows copying tracked state for members of reference types without hitting cycles.
Fixes #31395

### CODE REVIEW

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review.

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review.

**Comment by 333fred** [comment]
>         x.G.ToString(); // warning

Please number these

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:48408 in 91bbd39. [](commit_id = 91bbd39aa605fd45155c1aebfab835a8ec4c55c1, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -82247,6 +82235,69 @@ static void F(object o)
...
+class Program
+{
+    static void F()
+    {
+        var a1 = new A() { B = new B() };
+        var a2 = new A() { B = new B() { A = a1 } };
+        var a3 = new A() { B = new B() { A = a2 } };
+        a1.B.ToString();
+        a2.B.A.B.ToString();
+        a3.B.A.B.A.B.ToString();
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Consider putting a `// 1` on the warning lines here and in the below test.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -82247,6 +82235,69 @@ static void F(object o)
...
+        t.x1.x2.x3.y4.ToString();
+        t.x1.x2.x3.x4.y5.ToString();
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (10,9): warning CS8602: Dereference of a possibly null reference.
+                //         t.x1.x2.x3.x4.y5.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t.x1.x2.x3.x4.y5").WithLocation(10, 9));
+        }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >} [](start = 8, length = 1)

Consider adding tests we discussed if not already covered: `while (b) { x = x.F; }` and `while(b) { x.F = x; }`?

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: We have a number of similar cycle tests: see `Members_FieldCycle_*`. I believe those cases are covered.

---
In reply to: [284858771](https://github.com/dotnet/roslyn/pull/35720#discussion_r284858771) [](ancestors = 284858771)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Will add these comments in the next PR.

---
In reply to: [284816607](https://github.com/dotnet/roslyn/pull/35720#discussion_r284816607) [](ancestors = 284816607)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -318,7 +318,9 @@ protected override void Free()
             ArrayBuilder<(BoundReturnStatement, TypeWithAnnotations)> returnTypesOpt,
             VariableState initialState,
             Dictionary<BoundExpression, (NullabilityInfo, TypeSymbol)> analyzedNullabilityMapOpt)
-            : base(compilation, symbol, node, EmptyStructTypeCache.CreatePrecise(), trackUnassignments: true)
+            // Members of variables are tracked up to a fixed depth, to avoid cycles. The
+            // maxSlotDepth value is arbitrary but large enough to allow most scenarios.
+            : base(compilation, symbol, node, EmptyStructTypeCache.CreatePrecise(), trackUnassignments: true, maxSlotDepth: 5)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >maxSlotDepth: 5 [](start = 110, length = 15)

It looks like this change has far reaching impact that goes way beyond what the PR title says: "Copy nullable state of members unconditionally up to a certain depth." In particular, it looks like it limits tracking of everything to 5 levels of nesting in all scenarios. Even those that are not involved in copying states. Even for structure fields that never had problems with circularity and are handled by definite assignment check to any depth. I do not think this is the right thing to do.

Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Logged [https://github.com/dotnet/roslyn/pull/35720](https://github.com/dotnet/roslyn/pull/35720)

---
In reply to: [284911184](https://github.com/dotnet/roslyn/pull/35720#discussion_r284911184) [](ancestors = 284911184)

Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Looks like copy/paste error. You probably meant https://github.com/dotnet/roslyn/issues/35773

**Pull request merged**